R/update.R

```diff
@@ -308,8 +309,77 @@ setMethod("update", "parameter_list", function(object,.y,...) {
   object
 })
 
-##' @export
-##' @rdname update
-setMethod("update", "ev", function(object,y,...) {
-  
-})
+
+#' Update parameters, initials, and settings within a model object
+#' 
+#' The main use case for using [within] rather than [update] or [param] or
+#' [init] is when you want to update to a new value that is calculated from 
+#' the existing value.  See the example in details
+#' 
+#' Other model object slots that can be updated: `start`, `end`, `delta`, 
+#' `add`, `rtol`, `atol`, `hmax`, `maxsteps`.  These are include for convenience, 
+#' but we expect that most of the time these will get updated through the 
+#' update method.
+#' 
+#' @param data an object with class mrgmod
+#' @param expr expressions evaluated in an environment containing various model
+#' object components, including parameters, initial conditions, and others 
+#' (see details)
+#' @param ... not used
+#' 
+#' @examples
+#' mod <- mrgsolve:::house()
+#' 
+#' mod2 <- within(mod, {CL <- CL * 1.5})
+#' 
+#' mod$CL
+#' mod2$CL
+#' 
+#' @seealso [update]
+#' @name within
+#' @aliases within,mrgmod-method
+#' @md
+#' @export
+within.mrgmod <- function(data,expr,...) {
+  p <- as.list(param(data))
+  i <- as.list(init(data))
+  up <- list(
+    start = data@start, end = data@end, delta = data@delta, 
+    atol = data@atol, rtol = data@rtol, hmax = data@hmax,
+    maxsteps = data@maxsteps
+  )
+  up <- up[which(!(names(up) %in% c(names(p),names(i))))]
+  parent <- parent.frame()
+  env <- list2env(c(p,i,up),parent = parent)
+  eval(substitute(expr),env)
```

Comment:
whats the point of this eval here, it doesn't get saved anywhere? ... or does this evaluate in the environment? I'm not familiar with this code, but if I had to guess it would eval in the env, but just want to confirm.

Reply:
I think it evals in `env`;  I was following `base:::within.list`. 
```r
function (data, expr, keepAttrs = TRUE, ...) 
{
    parent <- parent.frame()
    e <- evalq(environment(), data, parent)
    eval(substitute(expr), e)
    if (keepAttrs) {
        l <- as.list(e, all.names = TRUE)
        del <- setdiff(names(data), (nl <- names(l)))
        data[nl] <- l
        data[del] <- NULL
        data
    }
    else {
        as.list(e, all.names = TRUE)
    }
}
``` 

I'm not totally sure about `evalq` then `eval`.  But I think setting `parent` in `list2env` does something along the same lines?  I can access a variable in the parent frame to use in `expr`

```r
x <- 2
within(mod, {
  CL = CL*x
})
```